File: packages/sdks/client-sdks/gatrix-js-client-sdk/src/WatchFlagGroup.ts
/**
 * WatchFlagGroup - Group multiple flag watchers for batch management
 * Provides convenience for registering/unregistering multiple watchers at once
 */
import { FeaturesClient } from './FeaturesClient';
import { FlagProxy } from './FlagProxy';

export class WatchFlagGroup {
  private client: FeaturesClient;
  private name: string;
  private unsubscribers: Array<() => void> = [];

  constructor(client: FeaturesClient, name: string) {
    this.client = client;
    this.name = name;
  }

  /**
   * Get the group name
   */
  getName(): string {
    return this.name;
  }

  /**
   * Watch a flag for realtime changes and add to this group.
   * Returns this group for chaining.
   */
  watchRealtimeFlag(flagName: string, callback: (flag: FlagProxy) => void | Promise<void>): this {
    const unsubscribe = this.client.watchRealtimeFlag(flagName, callback);
    this.unsubscribers.push(unsubscribe);
    return this;
  }

  /**
   * Watch a flag for realtime changes with initial state and add to this group.
   * Returns this group for chaining.
   */
  watchRealtimeFlagWithInitialState(
    flagName: string,
    callback: (flag: FlagProxy) => void | Promise<void>
  ): this {
    const unsubscribe = this.client.watchRealtimeFlagWithInitialState(flagName, callback);
    this.unsubscribers.push(unsubscribe);
    return this;
  }

  /**
   * Watch a flag for synced changes and add to this group.
   * Returns this group for chaining.
   */
  watchSyncedFlag(flagName: string, callback: (flag: FlagProxy) => void | Promise<void>): this {
    const unsubscribe = this.client.watchSyncedFlag(flagName, callback);
    this.unsubscribers.push(unsubscribe);
    return this;
  }

  /**
   * Watch a flag for synced changes with initial state and add to this group.
   * Returns this group for chaining.
   */
  watchSyncedFlagWithInitialState(
    flagName: string,
    callback: (flag: FlagProxy) => void | Promise<void>
  ): this {
    const unsubscribe = this.client.watchSyncedFlagWithInitialState(flagName, callback);
    this.unsubscribers.push(unsubscribe);
    return this;
  }

  /**
   * Unwatch all registered watchers in this group
   */
  unwatchAll(): void {
    for (const unsubscribe of this.unsubscribers) {
      unsubscribe();
    }
    this.unsubscribers = [];
  }

  /**
   * Alias for unwatchAll - destroys the group
   */
  destroy(): void {
    this.unwatchAll();
  }

  /**
   * Get the number of active watchers in this group
   */
  get size(): number {
    return this.unsubscribers.length;
  }
}

File: packages/sdks/client-sdks/gatrix-unity-client-sdk/src/Assets/Plugins/Gatrix/Runtime/WatchFlagGroup.cs
// WatchFlagGroup - Group multiple flag watchers for batch management

using System;
using System.Collections.Generic;

namespace Gatrix.Unity.SDK
{
    /// <summary>
    /// Batch management for multiple flag watchers.
    /// Supports chaining and bulk unsubscribe.
    /// </summary>
    public class WatchFlagGroup
    {
        private readonly FeaturesClient _client;
        private readonly string _name;
        private readonly List<Action> _unsubscribers = new List<Action>(4);

        public WatchFlagGroup(FeaturesClient client, string name)
        {
            _client = client;
            _name = name;
        }

        /// <summary>Get the group name</summary>
        public string GetName() => _name;

        /// <summary>Watch a flag for realtime changes and add to this group (returns this for chaining)</summary>
        public WatchFlagGroup WatchRealtimeFlag(string flagName, GatrixFlagWatchHandler callback)
        {
            var unsubscribe = _client.WatchRealtimeFlag(flagName, callback);
            _unsubscribers.Add(unsubscribe);
            return this;
        }

        /// <summary>Watch a flag for realtime changes with initial state and add to this group</summary>
        public WatchFlagGroup WatchRealtimeFlagWithInitialState(string flagName, GatrixFlagWatchHandler callback)
        {
            var unsubscribe = _client.WatchRealtimeFlagWithInitialState(flagName, callback);
            _unsubscribers.Add(unsubscribe);
            return this;
        }

        /// <summary>Watch a flag for synced changes and add to this group (returns this for chaining)</summary>
        public WatchFlagGroup WatchSyncedFlag(string flagName, GatrixFlagWatchHandler callback)
        {
            var unsubscribe = _client.WatchSyncedFlag(flagName, callback);
            _unsubscribers.Add(unsubscribe);
            return this;
        }

        /// <summary>Watch a flag for synced changes with initial state and add to this group</summary>
        public WatchFlagGroup WatchSyncedFlagWithInitialState(string flagName, GatrixFlagWatchHandler callback)
        {
            var unsubscribe = _client.WatchSyncedFlagWithInitialState(flagName, callback);
            _unsubscribers.Add(unsubscribe);
            return this;
        }

        /// <summary>Unwatch all registered watchers in this group</summary>
        public void UnwatchAll()
        {
            for (int i = 0; i < _unsubscribers.Count; i++)
            {
                _unsubscribers[i]();
            }
            _unsubscribers.Clear();
        }

        /// <summary>Alias for UnwatchAll - destroys the group</summary>
        public void Destroy() => UnwatchAll();

        /// <summary>Get the number of active watchers</summary>
        public int Size => _unsubscribers.Count;
    }
}
